File: solution/486. Predict the Winner/486_01.cpp
class Solution {
public:
    bool PredictTheWinner(vector<int>& nums) {
		vector<vector<int>> memo(nums.size(), vector<int> (nums.size()));
		return score(nums, 0, nums.size() - 1, memo) >= 0;
	}

	int score(vector<int>& nums, int s, int e, vector<vector<int>>& memo) {
		if (memo[s][e] != NULL) return memo[s][e];
		if (s == e) return nums[s];

		int a = nums[s] - score(nums, s + 1, e, memo);
		int b = nums[e] - score(nums, s, e - 1, memo);

		memo[s][e] = max(a, b);
		return memo[s][e];
	}
};

File: solution/486. Predict the Winner/486_01.py
class Solution:
    def PredictTheWinner(self, nums: List[int]) -> bool:
        @lru_cache
        def dp(i, j):
            if i == j:
                return nums[i]
            return max(nums[i] - dp(i + 1, j), nums[j] - dp(i, j - 1))
        return dp(0, len(nums) - 1) >= 0
